File: database-driver-api/src/main/java/de/lino/database/DatabaseRepository.java
package de.lino.database;

import de.lino.database.database.auth.Credentials;
import de.lino.database.database.DatabaseProvider;
import de.lino.database.database.DatabaseType;
import de.lino.database.utils.Pair;
import lombok.Getter;
import lombok.SneakyThrows;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.UnmodifiableView;

import java.util.List;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;

/**
 * Central entry point for managing the pool of registered {@link DatabaseProvider} instances.
 * <p>
 * A single {@link DatabaseRepository} instance is expected to exist per application and is
 * exposed through the generated {@code getInstance()} accessor once a concrete subclass has
 * installed itself via {@link #setInstance(DatabaseRepository)}. Implementations are responsible
 * for tracking
 * registered providers by numeric id, creating new providers for a given {@link DatabaseType}
 * and {@link Credentials}, and converting data between two registered providers.
 * <p>
 * Every synchronous operation declared here has a corresponding {@code *Async} default method
 * that executes the same logic on the common {@link CompletableFuture} pool.
 */
public abstract class DatabaseRepository {

    /**
     * The globally accessible instance of this repository, installed via
     * {@link #setInstance(DatabaseRepository)} and exposed through the generated
     * {@code getInstance()} accessor.
     */
    @Getter
    @Nullable
    private static DatabaseRepository instance;

    /**
     * Installs the given repository as the globally accessible instance returned by the
     * generated {@code getInstance()} accessor.
     *
     * @param instance the repository instance to install
     */
    protected static void setInstance(@NotNull DatabaseRepository instance) {
        DatabaseRepository.instance = instance;
    }

    /**
     * Get an unmodifiable list of all registered database.
     *
     * @return an unmodifiable {@code List} of every registered {@link DatabaseProvider}
     */
    @UnmodifiableView
    public abstract List<DatabaseProvider> getDatabaseProviderPool();

    /**
     * Get an unmodifiable list of all registered database by a specific type.
     *
     * @param databaseType the database type to filter by
     * @return an unmodifiable {@code List} of every registered {@link DatabaseProvider} of the
     * given type
     */
    @UnmodifiableView
    public abstract List<DatabaseProvider> getDatabaseProviderPool(@NotNull DatabaseType databaseType);

    /**
     * Shutdown all running database providers.
     */
    public abstract void shutdown();

    /**
     * Convert the content of a specific database to another one.
     *
     * @param sourceId Id of the database that shall be used as a resource database
     * @param targetId Id of the database that will be used as a destination database
     * @return a {@link Pair} of the two providers involved, the first one being the source, the
     * second one being the destination
     */
    public abstract Pair<DatabaseProvider, DatabaseProvider> convert(int sourceId, int targetId);

    /**
     * Get a specific database by id.
     *
     * @param id database id
     * @return an {@link Optional} containing the matching {@link DatabaseProvider}, or empty if
     * no database is registered under the given id
     */
    public abstract Optional<DatabaseProvider> findDatabaseProviderById(int id);

    /**
     * Register a new database.
     *
     * @param id           Id of the database
     * @param databaseType database type
     * @param credentials  login credentials
     * @return the newly created and registered {@link DatabaseProvider}
     */
    public abstract DatabaseProvider registerDatabaseProvider(int id, @NotNull DatabaseType databaseType, @NotNull Credentials credentials);

    /**
     * Shutdown a specific database and unregister it from the repository.
     *
     * @param id database id
     * @return the {@link DatabaseProvider} that was shut down and unregistered
     */
    public abstract DatabaseProvider unregisterDatabaseProvider(int id);

    /**
     * Execute the {@link #getDatabaseProviderPool()} process async.
     *
     * @return a {@link CompletableFuture} resolving to an unmodifiable list of every registered
     * {@link DatabaseProvider}
     */
    @UnmodifiableView
    public CompletableFuture<List<DatabaseProvider>> getDatabaseProviderPoolAsync() {
        return CompletableFuture.supplyAsync(this::getDatabaseProviderPool);
    }

    /**
     * Execute the {@link #getDatabaseProviderPool(DatabaseType)} process async.
     *
     * @param databaseType the database type to filter by
     * @return a {@link CompletableFuture} resolving to an unmodifiable list of every registered
     * {@link DatabaseProvider} of the given type
     */
    @UnmodifiableView
    public CompletableFuture<List<DatabaseProvider>> getDatabaseProviderPoolAsync(@NotNull DatabaseType databaseType) {
        return CompletableFuture.supplyAsync(() -> this.getDatabaseProviderPool(databaseType));
    }

    /**
     * Execute the {@link #shutdown()} process async.
     *
     * @return a {@link CompletableFuture} that completes once every database has been shut down
     */
    public CompletableFuture<Void> shutdownAsync() {
        return CompletableFuture.runAsync(this::shutdown);
    }

    /**
     * Execute the {@link #convert(int, int)} process async.
     *
     * @param sourceId Id of the database that shall be used as a resource database
     * @param targetId Id of the database that will be used as a destination database
     * @return a {@link CompletableFuture} resolving to a {@link Pair} of the source and
     * destination providers
     */
    public CompletableFuture<Pair<DatabaseProvider, DatabaseProvider>> convertAsync(int sourceId, int targetId) {
        return CompletableFuture.supplyAsync(() -> convert(sourceId, targetId));
    }

    /**
     * Execute the {@link #findDatabaseProviderById(int)} process async.
     *
     * @param id database id
     * @return a {@link CompletableFuture} resolving to an {@link Optional} containing the
     * matching {@link DatabaseProvider}, or empty if none is registered under the given id
     */
    @SneakyThrows
    public CompletableFuture<Optional<DatabaseProvider>> findDatabaseProviderByIdAsync(int id) {
        return CompletableFuture.supplyAsync(() -> findDatabaseProviderById(id));
    }

    /**
     * Execute the {@link #registerDatabaseProvider(int, DatabaseType, Credentials)} process
     * async.
     *
     * @param id           Id of the database
     * @param databaseType database type
     * @param credentials  login credentials
     * @return a {@link CompletableFuture} resolving to the newly created and registered
     * {@link DatabaseProvider}
     */
    public CompletableFuture<DatabaseProvider> registerDatabaseProviderAsync(int id, @NotNull DatabaseType databaseType, @NotNull Credentials credentials) {
        return CompletableFuture.supplyAsync(() -> registerDatabaseProvider(id, databaseType, credentials));
    }

    /**
     * Execute the {@link #unregisterDatabaseProvider(int)} process async.
     *
     * @param id database id
     * @return a {@link CompletableFuture} resolving to the {@link DatabaseProvider} that was shut
     * down and unregistered
     */
    public CompletableFuture<DatabaseProvider> unregisterDatabaseProviderAsync(int id) {
        return CompletableFuture.supplyAsync(() -> unregisterDatabaseProvider(id));
    }

}

File: python/database-driver-api/src/database_driver/api/database_repository.py
"""Mirror of ``de.lino.database.DatabaseRepository``."""

from __future__ import annotations

import asyncio
from abc import ABC, abstractmethod
from typing import ClassVar

from database_driver.api.database.auth.credentials import Credentials
from database_driver.api.database.database_provider import DatabaseProvider
from database_driver.api.database.database_type import DatabaseType
from database_driver.api.utils.pair import Pair


class DatabaseRepository(ABC):
    """Central entry point for managing the pool of registered ``DatabaseProvider``
    instances.

    A single ``DatabaseRepository`` instance is expected to exist per application and is
    exposed through :meth:`get_instance` once a concrete subclass has installed itself
    via :meth:`set_instance`. Implementations are responsible for tracking registered
    providers by numeric id, creating new providers for a given ``DatabaseType`` and
    ``Credentials``, and converting data between two registered providers.

    Every synchronous operation declared here has a corresponding ``*_async`` coroutine
    method that executes the same logic on a worker thread via ``asyncio.to_thread`` -
    the direct analogue of the Java edition's ``CompletableFuture`` default methods.
    """

    _instance: ClassVar[DatabaseRepository | None] = None

    @classmethod
    def get_instance(cls) -> DatabaseRepository | None:
        """Returns the globally accessible instance, or ``None`` if none was installed."""
        return DatabaseRepository._instance

    @classmethod
    def set_instance(cls, instance: DatabaseRepository) -> None:
        """Installs the given repository as the globally accessible instance returned by
        :meth:`get_instance`. Reserved for concrete subclasses, mirroring the Java
        edition's ``protected`` access."""
        DatabaseRepository._instance = instance

    @abstractmethod
    def get_database_provider_pool(self, database_type: DatabaseType | None = None) -> list[DatabaseProvider]:
        """Returns a list of all registered databases, optionally filtered to those of
        the given type. (The Java edition splits this into two overloads; Python folds
        them into one optional parameter.)"""

    @abstractmethod
    def shutdown(self) -> None:
        """Shuts down all running database providers."""

    @abstractmethod
    def convert(self, source_id: int, target_id: int) -> Pair[DatabaseProvider, DatabaseProvider]:
        """Converts the content of a specific database to another one.

        Args:
            source_id: Id of the database that shall be used as a resource database.
            target_id: Id of the database that will be used as a destination database.

        Returns:
            A :class:`Pair` of the two providers involved, the first one being the
            source, the second one being the destination.
        """

    @abstractmethod
    def find_database_provider_by_id(self, id: int) -> DatabaseProvider | None:
        """Returns the database registered under ``id``, or ``None`` if none is -
        standing in for the Java edition's ``Optional``."""

    @abstractmethod
    def register_database_provider(
        self, id: int, database_type: DatabaseType, credentials: Credentials
    ) -> DatabaseProvider:
        """Registers a new database.

        Args:
            id: Id of the database.
            database_type: Database type.
            credentials: Login credentials.

        Returns:
            The newly created and registered ``DatabaseProvider``.
        """

    @abstractmethod
    def unregister_database_provider(self, id: int) -> DatabaseProvider:
        """Shuts down a specific database and unregisters it from the repository.

        Returns:
            The ``DatabaseProvider`` that was shut down and unregistered.
        """

    # ------------------------------------------------------------------ async

    async def get_database_provider_pool_async(
        self, database_type: DatabaseType | None = None
    ) -> list[DatabaseProvider]:
        """Executes the :meth:`get_database_provider_pool` process async."""
        return await asyncio.to_thread(self.get_database_provider_pool, database_type)

    async def shutdown_async(self) -> None:
        """Executes the :meth:`shutdown` process async."""
        await asyncio.to_thread(self.shutdown)

    async def convert_async(self, source_id: int, target_id: int) -> Pair[DatabaseProvider, DatabaseProvider]:
        """Executes the :meth:`convert` process async."""
        return await asyncio.to_thread(self.convert, source_id, target_id)

    async def find_database_provider_by_id_async(self, id: int) -> DatabaseProvider | None:
        """Executes the :meth:`find_database_provider_by_id` process async."""
        return await asyncio.to_thread(self.find_database_provider_by_id, id)

    async def register_database_provider_async(
        self, id: int, database_type: DatabaseType, credentials: Credentials
    ) -> DatabaseProvider:
        """Executes the :meth:`register_database_provider` process async."""
        return await asyncio.to_thread(self.register_database_provider, id, database_type, credentials)

    async def unregister_database_provider_async(self, id: int) -> DatabaseProvider:
        """Executes the :meth:`unregister_database_provider` process async."""
        return await asyncio.to_thread(self.unregister_database_provider, id)
